horovod/run/runner.py

```diff
@@ -631,32 +651,74 @@ def _run(args):
 
     if args.run_func:
         # get the driver IPv4 address
-        driver_ip = network._get_driver_ip(nics)
+        driver_ip = network.get_driver_ip(nics)
         run_func_server = KVStoreServer(verbose=settings.verbose)
         run_func_server_port = run_func_server.start_server()
-        pickled_exec_func = cloudpickle.dumps(args.run_func)
         put_data_into_kvstore(driver_ip, run_func_server_port,
-                              'runfunc', 'func', pickled_exec_func)
+                              'runfunc', 'func', args.run_func)
 
         command = [sys.executable, '-m', 'horovod.run.run_task', str(driver_ip), str(run_func_server_port)]
 
         try:
-            _launch_job(args, remote_host_names, settings, nics, command)
+            _launch_job(args, settings, nics, command)
             results = [None] * args.np
             # TODO: make it parallel to improve performance
             for i in range(args.np):
-                pickled_result = read_data_from_kvstore(driver_ip, run_func_server_port,
-                                                        'runfunc_result', str(i))
-                results[i] = cloudpickle.loads(pickled_result)
+                results[i] = read_data_from_kvstore(driver_ip, run_func_server_port,
+                                                    'runfunc_result', str(i))
             return results
         finally:
             run_func_server.shutdown_server()
     else:
         command = args.command
-        _launch_job(args, remote_host_names, settings, nics, command)
+        _launch_job(args, settings, nics, command)
         return None
 
 
+def _run_elastic(args):
```

Comment:
In elastic, 'args.np` must be `> 1` when `args.max_np > args.np` and `args.nics is None`. Otherwise, the task-to-task address registration will be performed on a single host and `next_task_index = (index + 1) % num_hosts` makes the task connect to itself. This is fine in static mode because there will be no task-to-task communication, but in elastic mode with `args.max_np > args.np` there will be other tasks comming up later and then the detected common interfaces might be wrong then.

A simple exception when this `args` situation occurs should help to avoid this situation. Starting from a single worker with `args.max_np > args.np` is not very realistic anyway.

Reply:
I added some additional logic to `gloo_run.py` and `wait_for_available_hosts()` to address this.